test/subtype.jl

```diff
@@ -2107,8 +2107,33 @@ end
 
 let A = Tuple{typeof(identity), Type{Union{}}},
     B = Tuple{typeof(identity), typeof(Union{})}
-    @test A == B && (Base.isdispatchtuple(A) == Base.isdispatchtuple(B))
-end
+    @test A == B && (Base.isindivisibletype(A) == Base.isindivisibletype(B))
+end
+
+# isindivisibletype
+@test Base.isindivisibletype(Int)
+@test Base.isindivisibletype(Bool)
+@test Base.isindivisibletype(Float64)
+@test Base.isindivisibletype(Tuple{Float64})
+@test Base.isindivisibletype(Vector{Float64})
+@test !Base.isindivisibletype(Tuple{Union{Float64,Int64}})
+@test Base.isindivisibletype(Vector{Union{Float64,Int64}})
+@test !Base.isindivisibletype(Vector{T} where T <: Union{Float64,Int64})
+@test !Base.isindivisibletype(DataType)
+
+@test Base.isindivisibletype(Tuple{Tuple{Float64}})
+@test !Base.isindivisibletype(Tuple{Tuple{Union{Int,Float64}}})
+@test !Base.isindivisibletype(Tuple{Tuple})
+
+@test Base.isindivisibletype(Type{Int})
+@test Base.isindivisibletype(Type{Union{Int,Float64}})
+@test Base.isindivisibletype(Type{Any})
+
+@test Base.isdispatchtuple(Tuple{Tuple{Float64}})
+@test Base.isdispatchtuple(Tuple{Tuple{Type{Union{Int,Float64}}}})
+@test !Base.isdispatchtuple(Tuple{Tuple{Union{Int,Float64}}})
+@test !Base.isdispatchtuple(Tuple{Tuple{DataType}})
```

Comment:
These tests fail on master w/o this PR, which I assume is a bug:
```julia
@test Base.isdispatchtuple(Tuple{Tuple{Type{Union{Int,Float64}}}})
@test !Base.isdispatchtuple(Tuple{Tuple{DataType}})
```

Reply:
Those both return correct answers for me on master. What are you seeing:
```
julia> Base.isdispatchtuple(Tuple{Tuple{Type{Union{Int,Float64}}}})
false

julia> Base.isdispatchtuple(Tuple{Tuple{DataType}})
true
```

Reply:
`Tuple{Tuple{DataType}}` has a strict subtype (`Tuple{Tuple{Type{Int}}}`) that could appear in a call, making it not a dispatch tuple

`Tuple{Tuple{Type{Union{Int,Float64}}}}` has no such sub-types just like `Tuple{Type{Union{Int,Float64}}}` has none

Reply:
`Tuple{Type{Int}}` is not a constructible object, so cannot appear in a call

Reply:
Oh wow, that's... surprising. I didn't know that.

Anyway I think the definitions still coincide (`Tuple{Type{Int}}` is not inhabited so it does not affect divisibility either), so I'll update the computation to match

Reply:
Sounds good. I don't think our current definition is likely to remain valid for much longer (particularly the part about no subtypes or supertypes) so it is helpful to try to think about that we might replace it with. Probably at least helpful to clarify that "appear in call" means the same as "has_concrete_subtype"

Reply:
That does seem fatal actually...

This means that we have to refer to dispatch in this definition, since inhabited dispatch tuples != inhabited tuples (they are a strict super-set)